client/lib/signup/step-actions.js

```diff
@@ -150,6 +154,68 @@ function setThemeOnSite( callback, { siteSlug }, { themeSlug } ) {
 	} );
 }
 
+/**
+ * Gets username suggestions from the API.
+ *
+ * Ask the API to validate a username.
+ *
+ * If the API returns a suggestion, then the username is already taken.
+ * If there is no error from the API, then the username is free.
+ *
+ * @param {string} username The username to get suggestions for.
+ * @param {object} reduxState The Redux state object
+ */
+function getUsernameSuggestion( username, reduxState ) {
+	const fields = {
+		givesuggestions: 1,
+		username: username
+	};
+
+	// Clear out the local storage variable before sending the call.
+	reduxState.dispatch( {
+		type: SIGNUP_OPTIONAL_DEPENDENCY_SUGGESTED_USERNAME,
+		data: ''
+	} );
+
+	wpcom.undocumented().validateNewUser( fields, ( error, response ) => {
+		if ( error || ! response ) {
+			return null;
+		}
+
+		/**
+		 * Default the suggested username to `username` because if the validation succeeds would mean
+		 * that the username is free
+		 */
+		let resulting_username = username;
+
+		/**
+		 * Only start checking for suggested username if the API returns an error for the validation.
+		 */
+		if ( ! response.success ) {
+			const { messages } = response;
+
+			/**
+			 * The only case we want to update username field is when the username is already taken.
+			 *
+			 * This ensures that the validation is done
+			 *
+			 * Check for:
+			 *    - username taken error -
+			 *    - a valid suggested username
+			 */
+			if ( messages.username && messages.username.taken && messages.suggested_username ) {
+				resulting_username = messages.suggested_username.suggested_username;
```

Comment:
It's because of the way `WP_Error` handles adding additional data parameters to errors. I updated the API to return a not so redundant `suggested_username.data` instead.